File: somijeon/2022-06-16-3sum.js
// Given an integer array nums, return all the triplets [nums[i], nums[j], nums[k]] such that i != j, i != k, and j != k, and nums[i] + nums[j] + nums[k] == 0.

// Notice that the solution set must not contain duplicate triplets.

//* Example 1:

// Input: nums = [-1,0,1,2,-1,-4]
// Output: [[-1,-1,2],[-1,0,1]]

//* Example 2:

// Input: nums = [0]
// Output: []

const threeSum = function (nums) {
  const results = [];
  if (nums.length < 3) return results;
  nums = nums.sort((a, b) => a - b);

  for (let i = 0; i < nums.length - 2; i++) {
    // i = left
    if (nums[i] > 0) break;

    // i = left, skip numbers we've already seen
    if (i > 0 && nums[i] === nums[i - 1]) continue;

    let j = i + 1; // j = middle
    let k = nums.length - 1; // k = right

    while (j < k) {
      let sum = nums[i] + nums[j] + nums[k];
      if (sum === 0) {
        results.push([nums[i], nums[j], nums[k]]);
        while (nums[j] === nums[j + 1]) j++;
        while (nums[k] === nums[k - 1]) k--;
        j++;
        k--;
      } else if (sum < 0) {
        j++;
      } else {
        k--;
      }
    }
  }

  return results;
};

File: MinjiJu/2022-summer/2022-06-16-3sum.cpp
class Solution {
public:
    vector<vector<int>> threeSum(vector<int>& nums) {

        sort(nums.begin(), nums.end());
        vector<vector<int>> res;
        
        if(nums.size()<3) return res;
        
        for(int i=0; i<nums.size()-1; i++){
            
            if(i>0 && nums[i]==nums[i-1]) continue;
            
            int left=i+1, right=nums.size()-1;
            
            while(left<right){
                int sum = nums[i] + nums[left] + nums[right];
                if(sum<0) left++;
                else if(sum>0) right--;
                else{   // triplet found
                    res.push_back({nums[i],nums[left],nums[right]});
                    
                    while(left+1<right && nums[left]==nums[left+1]) left++;
                    while(left<right-1 && nums[right]==nums[right-1]) right--;
                    
                    left++;
                    right--;
                }
            }
        }
        return res;
    }
};